File: src/striatum/daemon_rpc/envelope.py
"""Envelope-v1 codec for daemon RPC."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping

from striatum.db import json_dumps, json_loads
from striatum.errors import StriatumError

SUPPORTED_ENVELOPE_VERSION = 1
DEFAULT_FRAMING = "json"


class RpcError(StriatumError):
    """Stable daemon RPC refusal."""

    def __init__(self, code: str, message: str, *, exit_code: int = 10, details: Mapping[str, Any] | None = None) -> None:
        super().__init__(message, exit_code=exit_code)
        self.code = code
        self.details = dict(details or {})


@dataclass(frozen=True)
class RpcEnvelope:
    schema_version: int
    request_id: str
    method: str
    params: dict[str, Any]
    capability_token: str | None = None
    deadline_ms: int = 0

    @classmethod
    def from_mapping(cls, payload: Mapping[str, Any]) -> "RpcEnvelope":
        version = payload.get("schema_version")
        if version != SUPPORTED_ENVELOPE_VERSION:
            raise RpcError(
                "version_incompatible",
                "daemon RPC envelope version is not supported",
                details={"supported_envelope": [SUPPORTED_ENVELOPE_VERSION], "received": version},
            )
        request_id = payload.get("request_id")
        method = payload.get("method")
        params = payload.get("params")
        if not isinstance(request_id, str) or not request_id:
            raise RpcError("schema_invalid", "daemon RPC request_id must be a non-empty string")
        if not isinstance(method, str) or "." not in method:
            raise RpcError("schema_invalid", "daemon RPC method must be a dotted string")
        if not isinstance(params, dict):
            raise RpcError("schema_invalid", "daemon RPC params must be a JSON object")
        token = payload.get("capability_token")
        if token is not None and not isinstance(token, str):
            raise RpcError("schema_invalid", "daemon RPC capability_token must be a string when present")
        deadline = payload.get("deadline_ms", 0)
        if not isinstance(deadline, int) or deadline < 0:
            raise RpcError("schema_invalid", "daemon RPC deadline_ms must be a non-negative integer")
        return cls(
            schema_version=SUPPORTED_ENVELOPE_VERSION,
            request_id=request_id,
            method=method,
            params=dict(params),
            capability_token=token,
            deadline_ms=deadline,
        )

    @classmethod
    def decode(cls, body: bytes | str) -> "RpcEnvelope":
        try:
            payload = json_loads(body.decode("utf-8") if isinstance(body, bytes) else body)
        except Exception as exc:  # noqa: BLE001 - malformed transport body.
            raise RpcError("schema_invalid", "daemon RPC body is not valid JSON") from exc
        if not isinstance(payload, dict):
            raise RpcError("schema_invalid", "daemon RPC body must be a JSON object")
        return cls.from_mapping(payload)

    def to_mapping(self) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "schema_version": self.schema_version,
            "request_id": self.request_id,
            "method": self.method,
            "params": self.params,
            "deadline_ms": self.deadline_ms,
        }
        if self.capability_token is not None:
            payload["capability_token"] = self.capability_token
        return payload

    def encode(self) -> bytes:
        return json_dumps(self.to_mapping()).encode("utf-8")


@dataclass(frozen=True)
class RpcResponse:
    schema_version: int
    request_id: str
    ok: bool
    data: dict[str, Any]
    audit_id: str | None = None

    @classmethod
    def ok_response(cls, *, request_id: str, data: Mapping[str, Any], audit_id: str | None = None) -> "RpcResponse":
        return cls(
            schema_version=SUPPORTED_ENVELOPE_VERSION,
            request_id=request_id,
            ok=True,
            data=dict(data),
            audit_id=audit_id,
        )

    @classmethod
    def error_response(cls, *, request_id: str, error: RpcError, audit_id: str | None = None) -> "RpcResponse":
        return cls(
            schema_version=SUPPORTED_ENVELOPE_VERSION,
            request_id=request_id,
            ok=False,
            data={"code": error.code, "message": str(error), "details": error.details},
            audit_id=audit_id,
        )

    def to_mapping(self) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "schema_version": self.schema_version,
            "request_id": self.request_id,
            "ok": self.ok,
            "data": self.data,
        }
        if self.audit_id is not None:
            payload["audit_id"] = self.audit_id
        return payload

    def encode(self) -> bytes:
        return json_dumps(self.to_mapping()).encode("utf-8")

File: go/pkg/rpc/envelope.go
package rpc

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
)

const (
	SupportedEnvelopeVersion = 1
	DefaultFraming           = "json"
)

type Error struct {
	Code    string         `json:"code"`
	Message string         `json:"message"`
	Details map[string]any `json:"details"`
	// Suggestion is the in-band remediation an agent can act on (RFC 0111
	// P2). Call sites may set it explicitly; when empty, ErrorResponse fills
	// the catalog's per-code default (error_catalog.go) centrally.
	Suggestion string `json:"suggestion,omitempty"`
	ExitCode   int    `json:"-"`
}

func (e *Error) Error() string {
	return e.Message
}

func NewError(code, message string, details map[string]any) *Error {
	return &Error{Code: code, Message: message, Details: details, ExitCode: 10}
}

type Envelope struct {
	SchemaVersion   int            `json:"schema_version"`
	RequestID       string         `json:"request_id"`
	Method          string         `json:"method"`
	Params          map[string]any `json:"params"`
	CapabilityToken string         `json:"capability_token,omitempty"`
	DeadlineMS      int            `json:"deadline_ms"`
}

type Response struct {
	SchemaVersion int            `json:"schema_version"`
	RequestID     string         `json:"request_id"`
	OK            bool           `json:"ok"`
	Data          map[string]any `json:"data"`
	AuditID       string         `json:"audit_id,omitempty"`
}

func DecodeEnvelope(body []byte) (Envelope, error) {
	dec := json.NewDecoder(bytes.NewReader(body))
	dec.UseNumber()
	var raw map[string]any
	if err := dec.Decode(&raw); err != nil {
		return Envelope{}, NewError("schema_invalid", "daemon RPC body is not valid JSON", nil)
	}
	if len(raw) == 0 {
		return Envelope{}, NewError("schema_invalid", "daemon RPC body must be a JSON object", nil)
	}
	return EnvelopeFromMap(raw)
}

func EnvelopeFromMap(payload map[string]any) (Envelope, error) {
	version, ok := intValue(payload["schema_version"])
	if !ok || version != SupportedEnvelopeVersion {
		return Envelope{}, NewError("version_incompatible", "daemon RPC envelope version is not supported", map[string]any{
			"supported_envelope": []int{SupportedEnvelopeVersion},
			"received":           payload["schema_version"],
		})
	}
	requestID, ok := payload["request_id"].(string)
	if !ok || requestID == "" {
		return Envelope{}, NewError("schema_invalid", "daemon RPC request_id must be a non-empty string", nil)
	}
	method, ok := payload["method"].(string)
	if !ok || method == "" {
		return Envelope{}, NewError("schema_invalid", "daemon RPC method must be a non-empty string", nil)
	}
	params, ok := payload["params"].(map[string]any)
	if !ok {
		return Envelope{}, NewError("schema_invalid", "daemon RPC params must be a JSON object", nil)
	}
	token := ""
	if value, exists := payload["capability_token"]; exists && value != nil {
		var tokenOK bool
		token, tokenOK = value.(string)
		if !tokenOK {
			return Envelope{}, NewError("schema_invalid", "daemon RPC capability_token must be a string when present", nil)
		}
	}
	deadline := 0
	if value, exists := payload["deadline_ms"]; exists {
		var deadlineOK bool
		deadline, deadlineOK = intValue(value)
		if !deadlineOK || deadline < 0 {
			return Envelope{}, NewError("schema_invalid", "daemon RPC deadline_ms must be a non-negative integer", nil)
		}
	}
	return Envelope{
		SchemaVersion:   SupportedEnvelopeVersion,
		RequestID:       requestID,
		Method:          method,
		Params:          params,
		CapabilityToken: token,
		DeadlineMS:      deadline,
	}, nil
}

func (e Envelope) Encode() ([]byte, error) {
	return json.Marshal(e)
}

func OKResponse(requestID string, data map[string]any, auditID string) Response {
	if data == nil {
		data = map[string]any{}
	}
	return Response{
		SchemaVersion: SupportedEnvelopeVersion,
		RequestID:     requestID,
		OK:            true,
		Data:          data,
		AuditID:       auditID,
	}
}

func ErrorResponse(requestID string, err error, auditID string) Response {
	rpcErr := &Error{}
	if !errors.As(err, &rpcErr) {
		rpcErr = NewError("internal_error", err.Error(), nil)
	}
	data := map[string]any{
		"code":    rpcErr.Code,
		"message": rpcErr.Message,
		"details": rpcErr.Details,
	}
	// RFC 0111 P2: fill the remediation centrally so the 165+ NewError call
	// sites stay untouched; an explicit call-site Suggestion wins over the
	// catalog default, and the key is omitted when neither exists.
	suggestion := rpcErr.Suggestion
	if suggestion == "" {
		suggestion = DefaultSuggestion(rpcErr.Code)
	}
	if suggestion != "" {
		data["suggestion"] = suggestion
	}
	return Response{
		SchemaVersion: SupportedEnvelopeVersion,
		RequestID:     requestID,
		OK:            false,
		Data:          data,
		AuditID:       auditID,
	}
}

func (r Response) Encode() ([]byte, error) {
	return json.Marshal(r)
}

func intValue(value any) (int, bool) {
	switch v := value.(type) {
	case int:
		return v, true
	case int64:
		return int(v), true
	case float64:
		if v == float64(int(v)) {
			return int(v), true
		}
	case json.Number:
		i, err := v.Int64()
		if err == nil {
			return int(i), true
		}
	}
	return 0, false
}

func DecodeResponse(body []byte) (Response, error) {
	var response Response
	if err := json.Unmarshal(body, &response); err != nil {
		return Response{}, fmt.Errorf("decode response: %w", err)
	}
	return response, nil
}
